extension/src/GameSingleton.cpp

```diff
@@ -9,6 +11,18 @@
 using namespace godot;
 using namespace OpenVic;
 
+TerrainVariant::TerrainVariant(std::string const& new_identfier,
+	colour_t new_colour, Ref<Image> const& new_image)
+	: HasIdentifier { new_identfier },
+	  HasColour { new_colour, true },
+	  image { new_image } {
+	assert(image.is_valid());
```

Comment:
Probably should be using Godot erroring instead of asserts.

Reply:
The actual null-checking is the responsibility of whoever calls the constructor (since constructors can't an return error value), if you look up `terrain_variants.add_item` there are 3 places it's used all with their own checks and custom error messages. The `assert` here and in similar constructors like `HasIdentifier` and `HasColour` are just to remind me of the conditions and catch silly mistakes. Still, the constructor shouldn't be publicly accessible, I'll re-commit with it private and `GameSingleton` added as a friend class.

Reply:
Well if you're gonna runtime asserts at all, you should be using Godot's erroring macros else if you miss a case it just crashes the entire engine.

Reply:
Re-committed without the assert, `TerrainVariant` itself would still be fine with an invalid reference, it would only become an issue when the terrain image stitching happens which is right next to the code which already does the actual invalid reference check.